metadata/metadata.go

```diff
@@ -195,12 +195,22 @@ func FromOutgoingContext(ctx context.Context) (MD, bool) {
 		return nil, false
 	}
 
-	mds := make([]MD, 0, len(raw.added)+1)
-	mds = append(mds, raw.md)
-	for _, vv := range raw.added {
-		mds = append(mds, Pairs(vv...))
+	out := raw.md.Copy()
+	for _, added := range raw.added {
+		if len(added)%2 == 1 {
+			panic(fmt.Sprintf("metadata: FromOutgoingContext got the odd number of input pairs for metadata: %d", len(added)))
+		}
+
+		var key string
+		for i, s := range added {
+			if i%2 == 0 {
+				key = strings.ToLower(s)
+				continue
+			}
+			out[key] = append(out[key], s)
+		}
```

Comment:
Would this be a little more optimal?  (It's certainly simpler.)

```go
for i := 0; i < len(added); i+=2 {
	key := strings.ToLower(added[i])
	out[key] = append(out[key], added[i+1])
}
```

Reply:
Done (had cargo culted from Pairs above; also updated).